clients/src/main/java/org/apache/kafka/clients/consumer/internals/Coordinator.java

```diff
@@ -460,6 +460,12 @@ private void maybeAutoCommitOffsetsSync() {
                     offsetAndMetadata.offset(), offsetAndMetadata.metadata()));
         }
 
+        // reset generation/consumerId to default values if using manual assignment
+        if (!subscriptions.partitionsAutoAssigned()) {
+            this.consumerId = OffsetCommitRequest.DEFAULT_CONSUMER_ID;
```

Comment:
I don't think an IllegalStateException gets thrown when empty lists come into the picture. Consider the following scenario:
```java
// assume polls may be interleaved
consumer.subscribe(Arrays.asList("topic"));
consumer.subscribe(new ArrayList<String>());
consumer.assign(Arrays.asList(new TopicPartition("topic", 0)));
consumer.assign(new ArrayList<TopicPartition>());
consumer.subscribe(Arrays.asList("topic"));
```